gt/main.c

```diff
@@ -870,14 +794,340 @@ print_unsent_policy(struct ggu_policy *policy)
+			if (ipv6_addrs_equal(ipv6_hdr->src_addr,
+					ggu_pkt->ipaddr.ip.v6.s6_addr))
+				return ggu_pkt;
+		}
+	}
+	return NULL;
+}
+
+static void
+prep_notify_pkt(struct ggu_notify_pkt *ggu_pkt)
+{
+	/*
+	 * Complete the packet fields that can only be done
+	 * when the packet is ready to be transmitted.
+	 */
+	struct udp_hdr *notify_udp;
+
+	if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *notify_ipv4 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv4_hdr *,
+				ggu_pkt->buf->l2_len);
+		notify_ipv4->total_length = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len);
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv4[1];
+		notify_udp->dgram_cksum =
+			rte_ipv4_phdr_cksum(notify_ipv4,
+				ggu_pkt->buf->ol_flags);
+	} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *notify_ipv6 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv6_hdr *, ggu_pkt->buf->l2_len);
+		notify_ipv6->payload_len = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len -
+			sizeof(struct ipv6_hdr));
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv6[1];
+		notify_udp->dgram_cksum =
+			rte_ipv6_phdr_cksum(notify_ipv6,
+				ggu_pkt->buf->ol_flags);
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u sending notification packet to Gatekeeper server with unknown IP version %hu\n",
+			rte_lcore_id(), ggu_pkt->ipaddr.proto);
+	}
+
+	notify_udp->dgram_len =
+		rte_cpu_to_be_16((uint16_t)(ggu_pkt->buf->data_len -
+			ggu_pkt->buf->l2_len - ggu_pkt->buf->l3_len));
+}
+
+static void
+send_notify_pkt(struct gt_config *gt_conf, struct gt_instance *instance,
+	struct ggu_notify_pkt *ggu_pkt)
+{
+	prep_notify_pkt(ggu_pkt);
+
+	if (rte_eth_tx_burst(gt_conf->net->front.id,
+			instance->tx_queue, &ggu_pkt->buf, 1) != 1) {
+		print_unsent_policies(ggu_pkt);
+		rte_pktmbuf_free(ggu_pkt->buf);
+	}
+
+	ggu_pkt->buf = NULL;
+	instance->num_ggu_pkts--;
+}
+
+/* Send all saved policy decision notification packets being buffered. */
+static void
+flush_notify_pkts(struct gt_config *gt_conf, struct gt_instance *instance)
+{
+	unsigned int max_pkts = gt_conf->max_ggu_notify_pkts;
+	struct rte_mbuf *bufs[max_pkts];
+	int num_to_send = 0;
+	int num_sent;
+	unsigned int i;
+
+	if (instance->ggu_pkts == NULL || instance->num_ggu_pkts == 0)
+		return;
+
+	for (i = 0; i < max_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
+			continue;
+
+		prep_notify_pkt(&instance->ggu_pkts[i]);
+		bufs[num_to_send++] = instance->ggu_pkts[i].buf;
+	}
+
+	num_sent = rte_eth_tx_burst(gt_conf->net->front.id,
+		instance->tx_queue, bufs, num_to_send);
+	if (unlikely(num_sent < num_to_send)) {
+		/*
+		 * Need to exclude the first @num_sent valid entries
+		 * found in instance->ggu_pkts from being printed.
+		 */
+		int sent_count = 0;
+		for (i = 0; i < max_pkts; i++) {
+			if (instance->ggu_pkts[i].buf == NULL)
```

Comment:
It'd be easier to iterate over `bufs`.

Reply:
I think it would be more complicated to iterate over `bufs`, since the mapping from `bufs` to `instance->ggu_pkts` is not 1:1.

`bufs` was created as an array of packet pointers to pass to the TX function, collected from the non-empty entries in `instance->ggu_pkts`. If we don't send all packets in `bufs`, then we need to find the entries in `instance->ggu_pkts` that correspond to the packets that were not sent in `bufs`.

The code I wrote above iterates over `instance->ggu_pkts` because it has to count through `num_sent` non-empty entries before it can start printing unsent policies in `instance->ggu_pkts`.

Reply:
The packets that were not sent are all packets starting at `bufs[num_sent]`. We've used analogous `for` loop every time `rte_eth_tx_burst()` is called with more than a packet in Gatekeeper.

Reply:
Ah, actually, I found where I got this assumption from:

        if (rte_pktmbuf_tailroom(ggu_pkt->buf) < GGU_MAX_DECISION_LEN)          
                send_notify_pkt(gt_conf, instance, ggu_pkt); 

If we are close to filling up a packet, we send it, meaning there can be gaps in the array `instance->ggu_pkts`. Since there can be gaps, the array might look like:

[x, a, x, x, b, c, x, x]

Which means bufs will look like:

[a.buf, b.buf, c.buf]

If `num_sent` is 2, then we need to iterate through `instance->ggu_pkts` because we need to count two valid entries, `a` and `b`, to correspond to the two sent packets from `bufs`. But then we need to print the unsent policies of `c`.

Reply:
In fact, the code is stronger than what you're assuming. For the example of `instance->ggu_pkts` that you gave, `bufs` would be [a, b, c]. So what I'm proposing will work anyway.

Reply:
Notice that there is a condition that may add a `NULL` between packets: when the packet that just received a new policy decision becomes full.

Reply:
What you just said about `num_sent` being 2 is exactly my point, only `c` must be printed and freed. The following loop after the whole `if` statement here will `NULL` the packet pointers.

Reply:
I just got your point, `print_unsent_policies()` needs `&instance->ggu_pkts[i]` not only the mbuf. In this case, we can combine the current `for` loop with the following `for` loop.

Reply:
OK, just combined these loops and rebased.